core/unit_test/default/TestDefaultDeviceDevelop.cpp

```diff
@@ -20,8 +20,109 @@
 
 #include <TestDefaultDeviceType_Category.hpp>
 
+namespace Kokkos {
+// Another attempt to insert dummy overload of parallel_for
+// Doesn't work: if the range_policy function returns different types
+// depending on the host vs device variant getting this failure:
+//   terminate called after throwing an instance of 'std::runtime_error'
+//     what():  cudaFuncGetAttributes(&attr, func) error(
+//     cudaErrorInvalidResourceHandle): invalid resource handle
+//     /ascldap/users/crtrott/Kokkos/kokkos/core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp:140
+
+#if 0
+  struct SinkPolicy {
+    template<class Arg>
+    KOKKOS_FUNCTION constexpr static bool sink(const Arg& arg) {
+      return sizeof(Arg) > 0 && (&arg!=nullptr);
+    }
+    template<class ... Args>
+    KOKKOS_FUNCTION explicit SinkPolicy(const Args& ... args) {
+      (void) (sink(args) && ... && true);
+    }
+  };
+
+  template<class ... Args>
+  KOKKOS_FUNCTION
+  void parallel_for(const SinkPolicy&, const Args& ...) {}
+#endif
+}  // namespace Kokkos
+
+namespace Kokkos {
+template <class Exec, class... Args>
+  requires(Kokkos::is_execution_space_v<Exec> &&
+           (!std::is_convertible_v<Args, size_t> && ... && true))
+KOKKOS_INLINE_FUNCTION auto range_policy(const Exec& exec, size_t N,
+                                         const Args&... args) {
+  // This constructor doesn't work on device
+  KOKKOS_IF_ON_HOST((return Kokkos::RangePolicy<Exec>(exec, 0, N, args...);))
+  // Calling device callable RangePolicy ctor to not get "calling host function"
+  // warning here If this guy doesn't return exactly the same type as above,
+  // getting issues
+  KOKKOS_IF_ON_DEVICE(
+      (return Kokkos::RangePolicy<Exec>(Impl::InternalTag{}, exec);))
+
+  // KOKKOS_IF_ON_DEVICE(( return SinkPolicy(args...); ))
+  //(void) exec; (void) N; (void) (SinkPolicy::sink(args) && ... && true);
+}
+
+template <class Exec, class... Args>
+  requires(!Kokkos::is_execution_space_v<Exec> &&
+           (!std::is_convertible_v<Args, size_t> && ... && true))
+KOKKOS_INLINE_FUNCTION auto range_policy(const Exec& exec, size_t N,
+                                         const Args&... args) {
+  return Kokkos::TeamVectorRange(exec, 0, N);
+}
+}  // namespace Kokkos
+
+template <class Exec, class X, class Y>
+KOKKOS_INLINE_FUNCTION void axpy(const Exec& exec, const X& x, const Y& y) {
+  Kokkos::parallel_for(
+      Kokkos::range_policy(exec, x.extent(0)),
+      KOKKOS_LAMBDA(const int& i) { x(i) += y(i); });
+}
```

Comment:
This is really the user code I want to achieve. Write generic Kokkos functions that can be called anywhere on the hierarchy.